libr/anal/block.c

```diff
@@ -893,7 +893,7 @@ R_API void r_anal_block_automerge(RList *blocks) {
 		RListIter *fit;
 		RAnalFunction *fcn;
 		r_list_foreach (block->fcns, fit, fcn) {
-			ht_up_insert (relevant_fcns, (ut64)fcn, NULL);
+			ht_up_insert (relevant_fcns, (ut64)(size_t)fcn, NULL);
```

Comment:
> Thats wrong. Theres ht_uu to store ut64-ut64 and ht_pp to store a couple of pointers which probably make more sense than storing pointers as ut64

Tried it, but imported some bugs. ht_pp won't do what you think. It assumes the key is string. See the comparator.
https://github.com/radareorg/radare2/blob/b4b48cbf86f99dd3952074c96660df926f81a5e3/shlr/sdb/src/ht_pp.c#L5-L17

Reply:
The comment itself makes clear why this hack is in there. but i see several things wrong:

`883         HtUP *relevant_fcns = ht_up_new0 (); // all the functions that contain some of our blocks (ht abused as a set)`

So this cod emust change and use `  7 SDB_API SetP *set_p_new(void) {` instead of `ht_up_new0`

or well.  aybe it would be better to use set_u_new() and hold the address of the function entrypoint instead of the pointer where the function structure is hold in memory. just in case this gets reallocated in the heap may result on glitchy results.

Also, the set api have no way to iterate over the elements, and this must be implemented to replace this in here

Reply:
Also, your highlight on pointing that the hashtable is missusing the `p`ointer as a `s`string must be fixed too.

Reply:
so as long as this refactoring is going further im fine for merging this and we can do the proper sdb fixing after this :) sgty?

Reply:
> so as long as this refactoring is going further im fine for merging this and we can do the proper sdb fixing after this :) sgty?

sgtm